### PR TITLE
`Assessment`: Remember collapsed instruction blocks

### LINKED ISSUE #5179: Remember collapsed instruction blocks when correcting exercises
### Is your feature request related to a problem?

When correcting exercises you often times collapse the top blocks of the instructions (problem statement, example solution) to be able to see the assessment insrutions clearly. Unfortunately after finishing an assessment and continuing with the next one, the same blocks are again extended. That means you have to collapse them for every new submission which can get annoying and is unnecessary since most of the time you only need the example solution or the problem statement if something is unclear about the assessment.

### Describe the solution you'd like

Remember which blocks were collapsed after a tutor finished an assessing an exercise. When the tutor clicks on assess next submission, Artemis allready collapses the blocks which were previously collapsed for the next submission so the tutor doesn't have to do it again.

### Describe alternatives you've considered

Make the problem statement and example solution block collapsed by default.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Dominik-Weinzierl**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.

### Motivation and Context

Closes #5179: Remember collapsed instructions blocks to increase the usability.

### Description

In order to remember the state, we store the state of the collapsed instruction block in the local storage.

### Steps for Testing

Prerequisites:
- 1 Tutor

1. Log in to Artemis
2. Assess an Exercise (open the window is enough)
3. Collapse one of the instructions blocks
4. Reload/Revisit the page
5. The block should be still collapsed (check local storage if needed)

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| expandable-section.component.ts | 85.71% | 100% |

### CODE REVIEW

**Review by alaraozdenler — COMMENTED** [review]
Tested on ts1. When I collapsed an instruction block and reloaded the page, it took me back to the course overview page. However, when I went to assess the same exercise again and click continue assessment, I saw that the instruction block stayed collapsed. I tried it with two different exercises and the same thing happened both times.

**Comment by Dominik-Weinzierl** [comment]
@alaraozdenler okay, I updated the testing steps accordingly :D but I understand correctly that it works as expected, right?

**Review by mallgaier — APPROVED** [review]
Tested on TS1, works as expected. Nice feature 👍

File: src/main/webapp/app/assessment/assessment-instructions/expandable-section/expandable-section.component.ts [hunk 1]
```diff
@@ -1,17 +1,34 @@
...
     // Icons
     faAngleRight = faAngleRight;
     faAngleDown = faAngleDown;
+
+    constructor(private localStorageService: LocalStorageService) {}
+
+    ngOnInit(): void {
+        const collapsed = this.localStorageService.retrieve(this.headerKey);
+        this.isCollapsed = collapsed !== undefined && collapsed !== null && (collapsed as boolean);
+        this.localStorageService.store(this.headerKey, this.isCollapsed);
```
Comment (Strohgelaender) [inline src/main/webapp/app/assessment/assessment-instructions/expandable-section/expandable-section.component.ts:24]: Can we maybe add some kind of prefix here? To indicate that this value saves the collapsed-value?

Comment (Dominik-Weinzierl) [reply src/main/webapp/app/assessment/assessment-instructions/expandable-section/expandable-section.component.ts]: Good idea :+1:

File: src/main/webapp/app/assessment/assessment-instructions/expandable-section/expandable-section.component.ts [hunk 2]
```diff
@@ -1,17 +1,43 @@
...
-    @Input() isCollapsed = false;
     @Input() hasTranslation = true;
     @Input() isSubHeader = false;
 
+    isCollapsed: boolean;
     // Icons
     faAngleRight = faAngleRight;
     faAngleDown = faAngleDown;
+
+    prefix = 'collapsed.';
```
Comment (Strohgelaender) [inline src/main/webapp/app/assessment/assessment-instructions/expandable-section/expandable-section.component.ts:19]:
```suggestion
    readonly prefix = 'collapsed.';
```

File: src/main/webapp/app/assessment/assessment-instructions/expandable-section/expandable-section.component.ts [hunk 3]
```diff
@@ -1,17 +1,43 @@
...
+     */
+    toggleCollapsed() {
+        this.isCollapsed = !this.isCollapsed;
+        this.localStorageService.store(this.key, this.isCollapsed);
+    }
+
+    /**
+     * Returns the key to identify the value in the local storage
+     */
+    get key() {
```
Comment (Strohgelaender) [inline src/main/webapp/app/assessment/assessment-instructions/expandable-section/expandable-section.component.ts:40]: Maybe rename to e.g. `storageKey`?

File: src/main/webapp/app/assessment/assessment-instructions/expandable-section/expandable-section.component.ts [hunk 4]
```diff
@@ -1,17 +1,43 @@
...
     faAngleRight = faAngleRight;
     faAngleDown = faAngleDown;
+
+    prefix = 'collapsed.';
+
+    constructor(private localStorageService: LocalStorageService) {}
+
+    ngOnInit(): void {
+        const collapsed = this.localStorageService.retrieve(this.key);
+        this.isCollapsed = collapsed !== undefined && collapsed !== null && (collapsed as boolean);
```
Comment (bassner) [inline src/main/webapp/app/assessment/assessment-instructions/expandable-section/expandable-section.component.ts:25]:
```suggestion
        this.isCollapsed = !!this.localStorageService.retrieve(this.key);
```

**Review by bassner — CHANGES_REQUESTED** [review]
One minor nitpick below :)

**Review by jbllhff — APPROVED** [review]
Tested on ts2, works great!

**Review by TheTrueVester — APPROVED** [review]
Tested on ts2. Everything worked as expected.

**Review by rhubinger — APPROVED** [review]
Tested on TS2, works as expected

**Comment by Dominik-Weinzierl** [comment]
@bassner @Strohgelaender Your suggestions are implemented in 78e51d6998164027ff4b1d5aba0752d6ad87edf8 :+1:

**Review by Strohgelaender — APPROVED** [review]
code looks good.

**Review by TheTrueVester — APPROVED** [review]
Retested on ts1. Everything worked as expected.

**Review by rhubinger — APPROVED** [review]
Reapprove due to the small code changes

**Pull request merged**